src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/AvoidPotentiallyExpensiveCallWhenLogging.cs

```diff
@@ -0,0 +1,333 @@
+
+                if (loggerMessageAttribute is null)
+                {
+                    return false;
+                }
+
+                // Try to get the log level from the attribute arguments.
+                logLevel = loggerMessageAttribute.NamedArguments
+                    .FirstOrDefault(p => p.Key.Equals(Level, StringComparison.Ordinal))
+                    .Value.Value as int?
+                    ?? LogLevelPassedAsParameter;
+
+                if (logLevel == LogLevelPassedAsParameter)
+                {
+                    logLevelArgumentOperation = invocation.Arguments
+                        .FirstOrDefault(a => a.Value.Type?.Name.Equals(LogLevel, StringComparison.Ordinal) ?? false);
+
+                    if (logLevelArgumentOperation is null)
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
+            }
+
+            public bool IsGuardedByIsEnabled(IOperation currentOperation, IInvocationOperation logInvocation, int logLevel, IArgumentOperation? logLevelArgumentOperation)
+            {
+                var conditional = currentOperation.GetAncestor<IConditionalOperation>(OperationKind.Conditional);
+
+                // This is the base case where no ancestor conditional is found.
+                if (conditional is null)
+                {
+                    return false;
+                }
+
+                var conditionInvocations = conditional.Condition
+                    .DescendantsAndSelf()
+                    .OfType<IInvocationOperation>();
+
+                // Check each invocation in the condition to see if it is a valid guard invocation, i.e. same instance and same log level.
+                // This is not perfect (e.g. 'if (logger.IsEnabled(LogLevel.Debug) || true)' is treated as guarded), but should be good enough to prevent false positives.
+                if (conditionInvocations.Any(IsValidIsEnabledGuardInvocation))
+                {
+                    return true;
+                }
+
+                // Recursively check the next conditional.
+                return IsGuardedByIsEnabled(conditional, logInvocation, logLevel, logLevelArgumentOperation);
+
+                bool IsValidIsEnabledGuardInvocation(IInvocationOperation invocation)
+                {
+                    if (!SymbolEqualityComparer.Default.Equals(_isEnabledMethod, invocation.TargetMethod) &&
+                        !invocation.TargetMethod.IsOverrideOrImplementationOfInterfaceMember(_isEnabledMethod))
+                    {
+                        return false;
+                    }
+
+                    return AreInvocationsOnSameInstance(logInvocation, invocation) && IsSameLogLevel(invocation.Arguments[0]);
+                }
+
+                static bool AreInvocationsOnSameInstance(IInvocationOperation invocation1, IInvocationOperation invocation2)
```

Comment:
A variation of this is used in other analyzers. I think this could be a good candidate for another extension method.